Arm/Memory/Separate.lean

```diff
@@ -115,6 +116,497 @@ example : mem_subset (BitVec.ofNat 64 (2^64 - 1)) 0#64 0#64 (BitVec.ofNat 64 (2^
+  constructor <;>
+    solve
+    | omega
+    | assumption
+
+-- theorem Nat.sub_mod_eq (x y : Nat) : (x - y) % t =
+
+private theorem Nat.sub_mod_eq_of_lt_of_le {x y : Nat} (hx : x < n) (hy : y ≤ x) :
+    (x - y) % n = (x % n) - (y % n) := by
+  rw [Nat.mod_eq_of_lt (by omega)]
+  rw [Nat.mod_eq_of_lt (by omega)]
+  rw [Nat.mod_eq_of_lt (by omega)]
+
+-- mem_subset' is a safe over-approximation of mem_subset.
+theorem mem_subset_of_mem_subset' (h : mem_subset' a an b bn) (han : an > 0) (hbn : bn > 0) :
+  mem_subset a (a + BitVec.ofNat 64 (an - 1)) b (b + BitVec.ofNat 64 (bn - 1)):= by
+  unfold mem_subset
+  obtain ⟨ha, hb, hstart, hend⟩ := h
+  unfold mem_legal' at ha hb
+  simp only [Nat.reducePow, Nat.add_one_sub_one, Bool.or_eq_true, decide_eq_true_eq,
+    Bool.and_eq_true]
+  by_cases hb : bn = 2^64
+  · left
+    apply BitVec.eq_of_toNat_eq
+    rw [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le ]
+    · rw [BitVec.toNat_add_eq_toNat_add_toNat]
+      · simp only [BitVec.toNat_ofNat, Nat.reducePow, Nat.reduceMod]
+        rw [Nat.mod_eq_of_lt (by omega)]
+        omega
+      · simp only [BitVec.toNat_ofNat, Nat.reducePow]
+        rw [Nat.mod_eq_of_lt (by omega)]
+        omega
+    · simp only [BitVec.le_def, BitVec.toNat_add, BitVec.toNat_ofNat, Nat.reducePow,
+      Nat.add_mod_mod]
+      rw [Nat.mod_eq_of_lt (by omega)]
+      omega
+  · have hbn : (BitVec.ofNat 64 (bn - 1)).toNat = (bn - 1) := by simp; omega
+    have han : (BitVec.ofNat 64 (an - 1)).toNat = (an - 1) := by simp; omega
+    rw [BitVec.le_def] at hstart
+    right
+    constructor
+    · rw [BitVec.add_sub_cancel_left]
+      · apply (BitVec.le_add_iff_sub_le _ _).mp
+        · rw [BitVec.le_def, BitVec.toNat_add, han, BitVec.toNat_add, hbn]; omega
+        . rw [BitVec.le_def, BitVec.toNat_add, han]; omega
+        · rw [hbn]; omega
+      · rw [hbn]; omega
+    · rw [BitVec.sub_le_sub_iff_right]
+      · rw [BitVec.le_def, BitVec.toNat_add, han]
+        omega
+      · assumption
+      · rw [BitVec.le_def, BitVec.toNat_add, han]
+        omega
+
+/- value of read_mem_bytes when separate. -/
+theorem read_mem_bytes_write_mem_bytes_eq_read_mem_bytes_of_mem_separate'
+  (hsep : mem_separate' x xn y yn) -- separation
+  (val : BitVec (yn * 8)) :
+    read_mem_bytes xn x (write_mem_bytes yn y val mem) =
+    read_mem_bytes xn x mem := by
+  simp only [Nat.reduceMul, memory_rules,
+    Nat.reduceAdd, BitVec.ofNat_eq_ofNat,
+    BitVec.cast_eq]
+  apply BitVec.eq_of_getLsb_eq
+  intros i
+  obtain ⟨hsrc, hdest, hsplit⟩ := hsep
+  rw [Memory.getLsb_read_bytes, Memory.getLsb_write_bytes, Memory.getLsb_read_bytes]
+  simp only [i.isLt]
+  simp only [decide_True, ite_eq_left_iff, Bool.true_and]
+  intros h₁
+  intros h₂
+  -- we need to make use of mem_separate to show that src_addr + i / 8 < dest_addr | src_addr + i/7 ≥ dest_addr + 16
+  exfalso
+  · rcases hsplit with this | this
+    · simp only [BitVec.not_lt, BitVec.le_def, BitVec.toNat_add,
```

Comment:
Try using `bitvec_rules` here?

Reply:
@shigoel This is too annoying to do throughout the file, as once in a while, `bitvec_rules` does not cover some lemma that `simp` does. I'd be happy to do this throughout the file, provided I'm allowed to take theorems that `simp` already has in its simp set which are missing from `bitvec_rules` / `minimal_theory`.

Example missing rules:
- `BitVec.not_lt`
- `BitVec.toNat_ofNat`
- `Fin.is_lt`
- `Bool.false_eq`
- `Bool.and_eq_false_imp`